modules/dnn/src/model.cpp

```diff
@@ -0,0 +1,194 @@
+
+Model::Model(const Net& network, const Size& size, const Scalar& mean, float scale, bool swapRB, bool crop)
+    : Net(network), impl_(new Impl{size, mean, scale, swapRB, crop}) {};
+
+Model& Model::setInputSize(const Size& size) {
+    impl_->size = size;
+    return *this;
+}
+
+Model& Model::setInputMean(const Scalar& mean) {
+    impl_->mean = mean;
+    return *this;
+}
+
+Model& Model::setInputScale(float scale) {
+    impl_->scale = scale;
+    return *this;
+}
+
+Model& Model::setInputCrop(bool crop) {
+    impl_->crop = crop;
+    return *this;
+}
+
+Model& Model::setInputSwapRB(bool swapRB) {
+    impl_->swapRB = swapRB;
+    return *this;
+}
+
+std::pair<int, float> Model::classify(InputArray frame) {
+    Mat blob;
+    blobFromImage(frame, blob, 1.0, impl_->size, Scalar(), impl_->swapRB, impl_->crop, CV_8U);
+    setInput(blob, "", impl_->scale, impl_->mean);
+    Mat out = forward();
+
+    double conf;
+    cv::Point maxLoc;
+    minMaxLoc(out, nullptr, &conf, nullptr, &maxLoc);
+
+    return {maxLoc.x, static_cast<float>(conf)};
+}
+
+void Model::detect(InputArray frame, CV_OUT std::vector<int>& classIds,
+                   CV_OUT std::vector<float>& confidences, CV_OUT std::vector<Rect2d>& boxes,
+                   float confThreshold, float nmsThreshold,  bool absoluteCoords)
+{
+    int frameWidth  = frame.cols();
+    int frameHeight = frame.rows();
+
+    if (impl_->size.width <= 0) impl_->size.width = frame.cols();
+    if (impl_->size.height <= 0) impl_->size.height = frame.rows();
+
+    Mat blob;
+    blobFromImage(frame, blob, 1.0, impl_->size, Scalar(), impl_->swapRB, impl_->crop, CV_8U);
+    setInput(blob, "", impl_->scale, impl_->mean);
+
+    if (getLayer(0)->outputNameToIndex("im_info") != -1) {  // Faster-RCNN or R-FCN
+        frameWidth  = impl_->size.width;
+        frameHeight = impl_->size.height;
+        Mat imInfo = (Mat_<float>(1, 3) << impl_->size.height, impl_->size.width, 1.6f);
+        setInput(imInfo, "im_info");
+    }
+
+    std::vector<Mat> detections;
+    std::vector<String> outNames = getUnconnectedOutLayersNames();
+    forward(detections, outNames);
+    CV_Assert(detections.size() > 0);
+
+    std::vector<String> layerNames = getLayerNames();
+    int lastLayerId = getLayerId(layerNames.back());
+    Ptr<Layer> lastLayer = getLayer(lastLayerId);
+
+    std::vector<int> predClassIds;
+    std::vector<Rect2d> predBoxes;
+    std::vector<float> predConf;
+
+    if (lastLayer->type == "DetectionOutput") {
+        for (int i = 0; i < detections.size(); ++i)
+        {
+            for (int j = 0; j < detections[i].size[2]; j++)
+            {
+                std::vector<Range> coord = { Range::all(), Range::all(), Range(j, j + 1), Range::all() };
+                Mat detect;
+                detections[i](coord).copyTo(detect);
+                std::vector<float> detection = detect.reshape(1, 1);
+                float conf = detection[2];
```

Comment:
Please use `float* data = detections[i].data` (as in https://github.com/opencv/opencv/blob/master/samples/dnn/object_detection.cpp sample). `std::vector<float> detection = detect.reshape(1, 1);` is an extra data copy.